launch/launch/utilities/signal_management.py

```diff
@@ -12,133 +12,157 @@
-        logger = launch.logging.get_logger(__name__)
-        logger.error("failed to set signal handlers in '{}'".format(__name__))
-        logger.error('this function must be called in the main thread')
-        raise
+    def __init__(
+        self,
+        loop: asyncio.AbstractEventLoop
+    ):
+        """
+        Instantiate manager.
+
+        :param loop: event loop that will handle the signals.
+        """
+        self.__loop = loop  # type: asyncio.AbstractEventLoop
+        self.__background_loop = None  # type: Optional[asyncio.AbstractEventLoop]
+        self.__handlers = {}  # type: dict
+        self.__prev_wakeup_handle = -1  # type: Union[int, socket.socket]
+        self.__wsock, self.__rsock = socket.socketpair()  # type: Tuple[socket.socket, socket.socket]  # noqa
+        self.__wsock.setblocking(False)
+        self.__rsock.setblocking(False)
+
+    def __enter__(self):
+        try:
+            self.__loop.add_reader(self.__rsock.fileno(), self.__handle_signal)
+        except NotImplementedError:
+            # Some event loops, like the asyncio.ProactorEventLoop
+            # on Windows, do not support asynchronous socket reads.
+            # So we emulate it.
+            self.__background_loop = asyncio.SelectorEventLoop()
+            self.__background_loop.add_reader(
+                self.__rsock.fileno(),
+                self.__loop.call_soon_threadsafe,
+                self.__handle_signal)
+
+            def run_background_loop():
+                asyncio.set_event_loop(self.__background_loop)
+                self.__background_loop.run_forever()
+
+            self.__background_thread = threading.Thread(target=run_background_loop)
+            self.__background_thread.start()
+        self.__prev_wakeup_handle = signal.set_wakeup_fd(self.__wsock.fileno())
+        if self.__prev_wakeup_handle != -1 and is_winsock_handle(self.__prev_wakeup_handle):
+            # On Windows, os.write will fail on a WinSock handle. There is no WinSock API
+            # in the standard library either. Thus we wrap it in a socket.socket instance.
+            self.__prev_wakeup_handle = socket.socket(fileno=self.__prev_wakeup_handle)
+        return self
+
+    def __exit__(self, type_, value, traceback):
+        if isinstance(self.__prev_wakeup_handle, socket.socket):
+            # Detach (Windows) socket and retrieve the raw OS handle.
+            prev_wakeup_handle = self.__prev_wakeup_handle.fileno()
+            self.__prev_wakeup_handle.detach()
+            self.__prev_wakeup_handle = prev_wakeup_handle
+        assert self.__wsock.fileno() == signal.set_wakeup_fd(self.__prev_wakeup_handle)
+        if self.__background_loop:
+            self.__background_loop.call_soon_threadsafe(self.__background_loop.stop)
+            self.__background_thread.join()
+            self.__background_loop.close()
+        else:
+            self.__loop.remove_reader(self.__rsock.fileno())
+
+    def __handle_signal(self):
```

Comment:
Now on Windows, this is being called from a different thread than the main event loop thread.
Are the callbacks we're registering thread safe? e.g.: is `LaunchService._shutdown()` thread safe? (which is the method called from most signal handlers we're registering).

Reply:
If you look at line 94 above, you'll that the background loop calls back the foreground loop. Or that was my intention.

Reply:
Makes sense, I didn't see that.